File: WebViewBrowser/app/src/main/java/com/noctusoft/webviewbrowser/VariablesManager.java
package com.noctusoft.webviewbrowser;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Manager class for user-defined variables that can be inserted into web forms.
 */
public class VariablesManager {
    private static final String TAG = "VariablesManager";
    private static final String SHARED_PREFS_NAME = "WebViewBrowserVariables";
    private static VariablesManager instance;

    private final SharedPreferences sharedPreferences;
    private Map<String, String> variables = new HashMap<>();

    /**
     * Gets the singleton instance of VariablesManager.
     *
     * @param context The application context.
     * @return The VariablesManager instance.
     */
    public static synchronized VariablesManager getInstance(Context context) {
        if (instance == null) {
            instance = new VariablesManager(context.getApplicationContext());
        }
        return instance;
    }

    private VariablesManager(Context context) {
        sharedPreferences = context.getSharedPreferences(SHARED_PREFS_NAME, Context.MODE_PRIVATE);
        loadVariables();
        setupDefaultVariables();
    }

    /**
     * Sets a variable value.
     *
     * @param name The variable name.
     * @param value The variable value.
     */
    public void setValue(String name, String value) {
        variables.put(name, value);
        saveVariables();
        Log.d(TAG, "Set variable: " + name + " = " + value);
    }

    /**
     * Gets a variable value by name.
     *
     * @param name The variable name.
     * @return The variable value, or null if not found.
     */
    public String getValue(String name) {
        return variables.get(name);
    }

    /**
     * Removes a variable.
     *
     * @param name The variable name to remove.
     */
    public void removeVariable(String name) {
        variables.remove(name);
        saveVariables();
        Log.d(TAG, "Removed variable: " + name);
    }

    /**
     * Gets all variable names.
     *
     * @return A list of all variable names.
     */
    public List<String> getAllVariableNames() {
        List<String> names = new ArrayList<>(variables.keySet());
        java.util.Collections.sort(names);
        return names;
    }

    /**
     * Gets all variables.
     *
     * @return A map of all variables.
     */
    public Map<String, String> getAllVariables() {
        return new HashMap<>(variables);
    }

    /**
     * Sets up default variables if none exist.
     */
    private void setupDefaultVariables() {
        if (variables.isEmpty()) {
            setValue("username", "user123");
            setValue("password", "password123");
            setValue("email", "john.doe@example.com");
            setValue("phone", "123-456-7890");
            Log.d(TAG, "Set up default variables");
        }
    }

    /**
     * Loads variables from SharedPreferences.
     */
    private void loadVariables() {
        try {
            Map<String, ?> allEntries = sharedPreferences.getAll();
            for (Map.Entry<String, ?> entry : allEntries.entrySet()) {
                if (entry.getValue() instanceof String) {
                    variables.put(entry.getKey(), (String) entry.getValue());
                }
            }
            Log.d(TAG, "Successfully loaded " + variables.size() + " variables");
        } catch (Exception e) {
            Log.e(TAG, "Error loading variables", e);
            variables = new HashMap<>();
        }
    }

    /**
     * Saves variables to SharedPreferences.
     */
    private void saveVariables() {
        try {
            SharedPreferences.Editor editor = sharedPreferences.edit();
            
            // Clear existing variables
            editor.clear();
            
            // Save all variables
            for (Map.Entry<String, String> entry : variables.entrySet()) {
                editor.putString(entry.getKey(), entry.getValue());
            }
            
            editor.apply();
            Log.d(TAG, "Successfully saved " + variables.size() + " variables");
        } catch (Exception e) {
            Log.e(TAG, "Error saving variables", e);
        }
    }
}

File: WebKitBrowser/Sources/VariablesManager.swift
import Foundation

/// Manager class for user-defined variables that can be inserted into web forms
class VariablesManager {
    
    // MARK: - Properties
    
    static let shared = VariablesManager()
    
    private var variables: [String: String] = [:]
    private let variablesFileName = "user_variables.plist"
    
    // MARK: - Initialization
    
    private init() {
        loadVariables()
        setupDefaultVariables()
    }
    
    // MARK: - Variable Management
    
    /// Set a variable value
    func setValue(_ value: String, for name: String) {
        variables[name] = value
        saveVariables()
    }
    
    /// Get a variable value by name
    func getValue(for name: String) -> String? {
        return variables[name]
    }
    
    /// Remove a variable
    func removeVariable(named name: String) {
        variables.removeValue(forKey: name)
        saveVariables()
    }
    
    /// Get all variable names
    func getAllVariableNames() -> [String] {
        return Array(variables.keys).sorted()
    }
    
    /// Get all variables
    func getAllVariables() -> [String: String] {
        return variables
    }
    
    // MARK: - Default Variables
    
    private func setupDefaultVariables() {
        // Only set default variables if none exist
        if variables.isEmpty {
            setValue("user123", for: "username")
            setValue("password123", for: "password")
            setValue("john.doe@example.com", for: "email")
            setValue("123-456-7890", for: "phone")
            print("Set up default variables")
        }
    }
    
    // MARK: - File Management
    
    private func loadVariables() {
        guard let documentsDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
            print("Error: Could not access documents directory")
            return
        }
        
        let fileURL = documentsDirectory.appendingPathComponent(variablesFileName)
        
        if FileManager.default.fileExists(atPath: fileURL.path) {
            do {
                let data = try Data(contentsOf: fileURL)
                if let loadedVariables = try PropertyListSerialization.propertyList(from: data, options: [], format: nil) as? [String: String] {
                    variables = loadedVariables
                    print("Successfully loaded \(variables.count) variables")
                }
            } catch {
                print("Error loading variables: \(error)")
            }
        } else {
            print("No variables file found, starting with defaults")
        }
    }
    
    private func saveVariables() {
        guard let documentsDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
            print("Error: Could not access documents directory")
            return
        }
        
        let fileURL = documentsDirectory.appendingPathComponent(variablesFileName)
        
        do {
            let data = try PropertyListSerialization.data(fromPropertyList: variables, format: .xml, options: 0)
            try data.write(to: fileURL)
            print("Successfully saved \(variables.count) variables")
        } catch {
            print("Error saving variables: \(error)")
        }
    }
}
